### DISCUSSION TITLE
TypeScript Autocomplete Configuration - NexusContext definition

### DISCUSSION THREAD

**Discussion by kervin5** [prisma/prisma#2707] [Q&A] [answered]
I'm trying to take advantage of VS and TypeScript autocomplete features. This used to work thanks to the use ObjectDefinitionBlock, but after migrating to the new nexus and nexus-plugin-prisma, VS and typescript complaint about **request** not existing in NexusContext. 

I'm adding an extra property to the context by using the below code and I know for sure that the property request actually exists in the context, it's just TypeScript that doesn't find it the definition.

```
import { schema } from 'nexus'
schema.addToContext(req => {
  return { ... }
})
```

For now I was able to disable the warnings by setting "strict": false, in my tsconfig.json and removing ObjectDefinitionBlock type, but I know this is not good and defeats the purpose of using TypeScript. Should I be using a different type instead of ObjectDefinitionBlock or how can I extend the definition and properties of NexusContext to get rid of the warnings? 

**Old code:**

```
import { ObjectDefinitionBlock } from '@nexus/schema/dist/definitions/objectType'
import { arg } from '@nexus/schema'

export default (t: ObjectDefinitionBlock<'Query'>) => {
  t.list.field('branches', {
    type: 'Branch',
    args: {
      where: arg({ type: 'BranchWhereInput' }),
    },
    resolve: async (parent, args, ctx) => {
      const user = await ctx.prisma.user.findOne({
        where: { id: ctx.request.user.id },
        include: { branch: { include: { company: true } } },
      })

      return ctx.prisma.branch.findMany({
        where: { company: { id: user?.branch?.company.id } },
      })
    },
  })
}
```

**New Code** to bypass the warnings and allow implicit "any"

```
import { schema } from 'nexus'

export default (t) => {
  t.list.field('branches', {
    type: 'Branch',
    args: {
      where: schema.arg({ type: 'BranchWhereInput' }),
    },
    resolve: async (parent, args, ctx) => {
      const user = await ctx.db.user.findOne({
        where: { id: ctx.request.user.id },
        include: { branch: { include: { company: true } } },
      })

      return ctx.db.branch.findMany({
        where: { company: { id: user?.branch?.company.id } },
      })
    },
  })
}

```

**Comment by ryands17** [comment]
Hey @kervin5 :wave: 

> I'm trying to take advantage of VS and TypeScript autocomplete features. This used to work thanks to the use ObjectDefinitionBlock, but after migrating to the new nexus and nexus-plugin-prisma, VS and typescript complaint about **request** not existing in NexusContext.
> 
> I'm adding an extra property to the context by using the below code and I know for sure that the property request actually exists in the context, it's just TypeScript that doesn't find it the definition.
> 
> ```
> import { schema } from 'nexus'
> schema.addToContext(req => {
>   return { ... }
> })
> ```

Regarding the above, `addToContext` automatically adds typings based on the return type of the context, so that should work out of the box without any changes.

> **Old code:**
> 
> ```
> import { ObjectDefinitionBlock } from '@nexus/schema/dist/definitions/objectType'
> import { arg } from '@nexus/schema'
> 
> export default (t: ObjectDefinitionBlock<'Query'>) => {
>   t.list.field('branches', {
>     type: 'Branch',
>     args: {
>       where: arg({ type: 'BranchWhereInput' }),
>     },
>     resolve: async (parent, args, ctx) => {
>       const user = await ctx.prisma.user.findOne({
>         where: { id: ctx.request.user.id },
>         include: { branch: { include: { company: true } } },
>       })
> 
>       return ctx.prisma.branch.findMany({
>         where: { company: { id: user?.branch?.company.id } },
>       })
>     },
>   })
> }
> ```
> 
> **New Code** to bypass the warnings and allow implicit "any"
> 
> ```
> import { schema } from 'nexus'
> 
> export default (t) => {
>   t.list.field('branches', {
>     type: 'Branch',
>     args: {
>       where: schema.arg({ type: 'BranchWhereInput' }),
>     },
>     resolve: async (parent, args, ctx) => {
>       const user = await ctx.db.user.findOne({
>         where: { id: ctx.request.user.id },
>         include: { branch: { include: { company: true } } },
>       })
> 
>       return ctx.db.branch.findMany({
>         where: { company: { id: user?.branch?.company.id } },
>       })
>     },
>   })
> }
> ```

As for this, you can add the exports in the following manner to obtain :

```ts
import { core } from 'nexus/components/schema'
import { schema } from 'nexus'

export default (t: core.ObjectDefinitionBlock<'Query'>) => {
  t.list.field('branches', {
    type: 'Branch',
    args: {
      where: schema.arg({ type: 'BranchWhereInput' }),
    },
    resolve: async (parent, args, ctx) => {
      const user = await ctx.db.user.findOne({
        where: { id: ctx.request.user.id },
        include: { branch: { include: { company: true } } },
      })

      return ctx.db.branch.findMany({
        where: { company: { id: user?.branch?.company.id } },
      })
    },
  })
}
```

Let me know if this works for you :)

**Reply by ryands17** [reply]
Hey @kervin5 did the above change work for you?

**Reply by kervin5** [reply]
Hi @ryands17, thank you for your help here and in the chat. I imported the type from core and enabled back strict to "true", but unfortunately I still get the warning about **request** not existing in the context. 

<img width="624" alt="image" src="https://user-images.githubusercontent.com/15281277/84315022-e144b880-ab1d-11ea-87e6-fe4e8426754e.png">

If I run **nexus build**, I also get a similar message.

<img width="786" alt="image" src="https://user-images.githubusercontent.com/15281277/84315293-51ebd500-ab1e-11ea-831d-710e20dd6d1a.png">

This is my current tsconfig.json

```json
{
  "compilerOptions": {
    "rootDir": ".",
    "lib": ["esnext", "DOM"],
    "esModuleInterop": true,
    "strict": true,
    "target": "ES2018",
    "moduleResolution": "node",
    "plugins": [{ "name": "nexus/typescript-language-service" }]
  },
  "include": ["."]
}
```

and this is how I'm adding request and response to the context

```typescript
 server.express.use((request: Request, response: any, next: any) => {
    schema.addToContext((req) => {
      return { request: request, response: response }
    })
    next()
  })
```

**Reply by ryands17** [reply, marked as answer]
@kervin5 Could you try writing it like this?

```ts
schema.extendType({
    type: 'Query',
    definition(t) {
       t.list.field('branches', {
         type: 'Branch',
         args: {
           where: schema.arg({ type: 'BranchWhereInput' }),
        },
      resolve: async (parent, args, ctx) => {
      const user = await ctx.db.user.findOne({
        where: { id: ctx.request.user.id },
        include: { branch: { include: { company: true } } },
      })

      return ctx.db.branch.findMany({
        where: { company: { id: user?.branch?.company.id } },
      })
    },
  })
    },
});
```

I think writing it outside as a standalone method is causing the Context to not be picked up from the types.

Let me know if this works.

**Reply by exactstaff** [reply]
@ryands17 I really appreciate your help.

Both approaches that you provided me with worked when I made some changes to the way that I was adding the properties to the context. 

Before I was importing schema on a separate file and calling **schema.addToContext** inside a function to inject the properties that I needed. Now, the autocomplete started working after doing something like this in my entry point.

```typescript
schema.addToContext((req) => {
  return { ...expressContext }
})

use(prisma({ features: { crud: true } }))
use(permissions)
```

**Comment by ludralph** [comment]
Hi @kervin5,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
